File: sdk/typescript/src/context.ts
/**
 * Context propagation for pipeline ID and stage using AsyncLocalStorage.
 *
 * Allows setting pipeline-level metadata that automatically propagates
 * through async call chains without explicit passing.
 */

import { AsyncLocalStorage } from "node:async_hooks";

export interface HikariContext {
  pipelineId?: string;
  stage?: string;
}

const storage = new AsyncLocalStorage<HikariContext>();

/**
 * Get the current pipeline ID from async context.
 *
 * @returns The pipeline ID, or undefined if not set
 */
export function getPipelineId(): string | undefined {
  return storage.getStore()?.pipelineId;
}

/**
 * Set the pipeline ID in the current async context.
 *
 * @param pipelineId - The pipeline identifier to set
 */
export function setPipelineId(pipelineId: string): void {
  const store = storage.getStore();
  if (store) {
    store.pipelineId = pipelineId;
  } else {
    storage.enterWith({ pipelineId });
  }
}

/**
 * Get the current stage name from async context.
 *
 * @returns The stage name, or undefined if not set
 */
export function getStage(): string | undefined {
  return storage.getStore()?.stage;
}

/**
 * Set the stage name in the current async context.
 *
 * @param stage - The stage name to set
 */
export function setStage(stage: string): void {
  const store = storage.getStore();
  if (store) {
    store.stage = stage;
  } else {
    storage.enterWith({ stage });
  }
}

/**
 * Run a function with a specific Hikari context.
 *
 * @param context - The context to set for the function execution
 * @param fn - The function to run with the given context
 * @returns The result of the function
 */
export function runWithContext<T>(
  context: HikariContext,
  fn: () => T
): T {
  return storage.run(context, fn);
}

File: sdk/python/src/hikari/context.py
"""Context propagation helpers using ``contextvars``.

Provides per-async-task storage of ``pipeline_id`` and ``stage`` so that
provider patches can read them without explicit parameter threading.
"""

from __future__ import annotations

import contextvars

_pipeline_id: contextvars.ContextVar[str | None] = contextvars.ContextVar(
    "hikari_pipeline_id", default=None
)
_stage: contextvars.ContextVar[str | None] = contextvars.ContextVar(
    "hikari_stage", default=None
)


def get_pipeline_id() -> str | None:
    """Return the current pipeline ID, or ``None`` if not set."""
    return _pipeline_id.get()


def set_pipeline_id(pipeline_id: str) -> None:
    """Set the pipeline ID on the current async context."""
    _pipeline_id.set(pipeline_id)


def get_stage() -> str | None:
    """Return the current stage override, or ``None`` if not set."""
    return _stage.get()


def set_stage(stage: str) -> None:
    """Override the auto-derived stage name on the current async context."""
    _stage.set(stage)
